node/src/components/gossiper/gossip_table.rs

```diff
@@ -209,25 +209,25 @@ impl<T: Copy + Eq + Hash + Display> GossipTable<T> {
             return GossipAction::Noop;
         }
 
-        if let Some(state) = self.paused.get_mut(data_id) {
-            let _ = state.holders.insert(holder);
-            return GossipAction::Noop;
+        let update = |state: &mut State| {
+            let _ = state.holders.insert(holder.clone());
+        };
+
+        if let Some(action) = self.update_current(data_id, update) {
+            return action;
         }
 
-        match self.current.entry(*data_id) {
-            Entry::Occupied(mut entry) => {
-                let is_new = false;
-                let state = entry.get_mut();
-                let _ = state.holders.insert(holder);
-                state.action(self.infection_target, self.holders_limit, is_new)
-            }
-            Entry::Vacant(entry) => {
-                let is_new = true;
-                let state = entry.insert(State::default());
-                let _ = state.holders.insert(holder);
-                state.action(self.infection_target, self.holders_limit, is_new)
-            }
+        if let Some(action) = self.update_paused(data_id, update) {
+            return action;
         }
+
+        // This isn't in finished, current, or paused - add a new entry to current.
+        let mut state = State::default();
+        update(&mut state);
+        let is_new = true;
+        let action = state.action(self.infection_target, self.holders_limit, is_new);
+        let _ = self.current.insert(*data_id, state);
+        action
```

Comment:
Looks like this could be extracted to a method `insert_new_item(data_id) -> Action`

Reply:
I was in two minds about it myself, but we'd also need to pass in the `update` closure or similar since we don't do the same thing when calling this inside `new_partial_data` vs `new_complete_data` (`new_complete_data` also sets `state.held_by_us = true;`)

I do like DRY, but in this case I think it'd actually make the code _less_ readable.  If you feel strongly though, I'll make the change - I'm not too fussed either way.

Reply:
if we pass in `holder: Option<NodeId>`, this should be enough to figure out if the data comes from a client and is held by us.